Lab/NB7 extension compatibility docs (#14838)
jupyterlab/jupyterlab: docs/source/extension/extension_multiple_ui.rst

.. Copyright (c) Jupyter Development Team.
.. Distributed under the terms of the Modified BSD License.

.. _multiple_ui_extensions:

> Extensions Targeting Multiple Applications
> ==========================================

Reviewer: This is kind of long. I  think we can remove "Extensions" because this doc is in "Develop Extensions" chapter.

```suggestion
Targeting Multiple Applications
===============================
```

Do you feel strongly about keeping "Extensions" part in?

![image](https://github.com/jupyterlab/jupyterlab/assets/5832902/fc791395-8f5f-48f4-a02c-ae260fecec89)

> *Make your extensions work in JupyterLab 4, Notebook 7 and more!*

Reviewer: ```suggestion
*Make your extensions work in JupyterLab, Notebook 7+ and more!*
```

Jupyter Notebook 7 is built with components from JupyterLab, and since
both use the same building blocks, that means your extension can work
on both (or any other frontend built with JupyterLab components) with
little or no modification depending on its design.

This guide will give you an overview of compatibility features, then a
tutorial and reference code covering some of the topics mentioned here.
If you don't know how to make extensions, you can read more about the
basics at :ref:`the Extension Tutorial <extension_tutorial>` or the
:ref:`the Extensions page <extension_dev>`.

> How Compatibility Works
> -----------------------
> 
> At a high level, extensions for JupyterLab and Jupyter Notebook both
> typically start from a template project, which you can download and set up
> using instructions from `the extension tutorial <https://jupyterlab.readthedocs.io/en/latest/extension/extension_tutorial.html>`_).
> Once your template is ready, you can start adding components and features to build your extension.
> 

Reviewer: This paragraph does not answer the question yet

```suggestion
Getting started
---------------

At a high level, extensions for JupyterLab and Jupyter Notebook both
typically start from a template project, which you can download and set up
using instructions from `the extension tutorial <https://jupyterlab.readthedocs.io/en/latest/extension/extension_tutorial.html>`_).
Once your template is ready, you can start adding components and features to build your extension.

How Compatibility Works
-----------------------

```

Reviewer: + we should use sphinx references here, not fully specified links

> An extension for JupyterLab (and for Notebook 7) is made up of a `series <https://jupyterlab.readthedocs.io/en/latest/extension/extension_dev.html>`_

Reviewer: Should use sphinx reference, not a fully specified link

of bundled `plugins <https://lumino.readthedocs.io/en/latest/api/interfaces/application.IPlugin.html#requires>`_,
and those plugins typically use components from the interface toolkit `Lumino <https://lumino.readthedocs.io/en/latest/api/index.html>`_
as well as the `JupyterLab API <https://jupyterlab.readthedocs.io/en/latest/api/index.html>`_

> (among others) to help build your extension's look and behavior (both are
> written in Typescript).

Reviewer: Both is ambig

```suggestion
(among others) to help build your extension's look and behavior. Both
Lumino and JupyterLab API are written in Typescript.
```

This is how basic compatibility features work: both apps use the same building
blocks and methods. For instance, both JupyterLab and Notebook 7 accept Lumino widgets

> as interface components, both apps allow you to add them to the interface by

Reviewer: ```suggestion
as interface components, both apps enable you to add them to the interface by
```

specifying an "area" to place them into, and extensions for both use the same
basic ``JupyterFrontendPlugin`` class.

How to Achieve Compatibility
----------------------------

> Compatibility can be achieved with literally no effort in some simple

Reviewer: ```suggestion
Compatibility can be achieved without additional effort in some simple
```

or with "little effort". "literally no effort" will be frustrating when it does not work for the user (see https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#writing-style)

pattern), and it is what enables the compatibility solutions discussed here.

Each plugin uses some properties (the ``requires`` and ``optional`` properties) to
request features it wants which are provided by other plugins that have been
loaded into JupyterLab. When your plugin requests features, the system sends
them to your plugin's activate function if they are available.

You can build compatible extensions by taking advantage of these plugin
properties, and how the plugin system uses them:

- When you designate a feature in the ``requires`` list of your
  plugin, JupyterLab will only load your plugin if that feature is available.
- By designating a feature in the ``optional`` list, JupyterLab will pass you
  an object for it (if it's available) or ``null`` if it's not.

So, these capabilities form the backbone of extension compatibility. You can
use them to make checks in your extensions that will allow them to function in
both JupyterLab and Jupyter Notebook 7 (and others).

JupyterLab itself is a :ref:`provider <dependency-injection-basic-info>` of many features through its built-in plugins,

> which you can read more about in the `Common Extension Points document <https://jupyterlab.readthedocs.io/en/latest/extension/extension_points.html>`_.

Reviewer: Please change to a shpinx reference instead of fully specified link

    if (statusBar) {
      const statusBarWidget = new ShoutStatusBarSummary();

      statusBar.registerStatusItem('shoutStatusBarSummary', {
        item: statusBarWidget
      });

      // Connect to the messageShouted to be notified when a new message
      // is published and react to it by updating the status bar widget.
      shoutWidget.messageShouted.connect((widget: ShoutWidget, time: Date) => {
        statusBarWidget.setSummary(
          'Last Shout: ' + widget.lastShoutTime?.toString() ?? '(None)'
        );
      });
    }

Using Required Features to Switch Behaviors
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Another pattern you can follow is to export a list of plugins from your
extension, then use different "requires" features to select different
behaviors based on which app the extension is currently running in.

Here is a snippet from `this sample extension <https://github.com/jupyterlab/extension-examples/tree/main/clap-button-message>`_
which adds a *clap* button to the top area in JupyterLab, or to the
right sidebar in Jupyter Notebook 7 (you can read the full extension
example code there):

.. code:: TypeScript

    /**

>      * Initialization data for the @jupyterlab-examples/clap-button JupyterLab extension.

Reviewer: This code snippets do not read great on more narrow screens, with some strings and comments requiring scrolling. Here is one case that might be easy cut:

```suggestion
     * Data for the @jupyterlab-examples/clap-button JupyterLab plugin.
```

Note `plugin` is more correct here than `extension` in technical sense.

     */
    const pluginJupyterLab: JupyterFrontEndPlugin<void> = {
      id: '@jupyterlab-examples/clap-button:pluginLab',
      description: 'Adds a clap button to the top area JupyterLab',
      autoStart: true,
      requires: [ILabShell],
      activate: (app: JupyterFrontEnd, labShell: ILabShell) => {
        console.log(
          'JupyterLab extension @jupyterlab-examples/clap-button is activated!'
        );

        // Create a ClapWidget and add it to the interface in the top area
        const clapWidget = new ClapWidget();

>         clapWidget.id = 'JupyterLabClapWidgetLab'; // Widgets need an id

Reviewer: I would keep it only for the first time (above) to reduce noise - this tutorial is not about widgets.

```suggestion
        clapWidget.id = 'JupyterLabClapWidgetLab';
```

        app.shell.add(clapWidget, 'top');
      }
    };

    /**

>      * Initialization data for the @jupyterlab-examples/clap-button Jupyter Notebook extension.

Reviewer: ```suggestion
     * Data for the @jupyterlab-examples/clap-button Jupyter Notebook plugin.
```

     */
    const pluginJupyterNotebook: JupyterFrontEndPlugin<void> = {
      id: '@jupyterlab-examples/clap-button:pluginNotebook',
      description: 'Adds a clap button to the right sidebar of Jupyter Notebook 7',
      autoStart: true,
      requires: [INotebookShell],
      activate: (app: JupyterFrontEnd, notebookShell: INotebookShell) => {
        console.log(
          'Jupyter Notebook extension @jupyterlab-examples/clap-button is activated!'
        );

        // Create a ClapWidget and add it to the interface in the right area
        const clapWidget = new ClapWidget();

>         clapWidget.id = 'JupyterNotebookClapWidgetNotebook'; // Widgets need an id

Reviewer: Same as above

```suggestion
        clapWidget.id = 'JupyterNotebookClapWidgetNotebook';
```
